HDDS-5490. Remove an OM node from HA ring

hanishakoneru opened the pull request:
## What changes were proposed in this pull request?

Adds support to remove an OM node from an HA cluster.

To remove an OM, we need to follow the below steps:
1. Update ozone-site.xml of each OM node by adding the to be decommissioned nodeId to the config `ozone.om.decommissioned.nodes.<om-service-id>`.
2. Run the following command:
`ozone admin om decommission -id=<om-service-id> -nodeid=<decommissioned-node-id> -hostname=<decommissioned-node-hostname [(optional) force]`

The DecommssionOM subcommand will verify that all the existing OMs configs have been updated with the decommissioned node. If the verification is successful, it will send a decommission request via OMAdmin protocol to the leader OM. The Leader OM will submit a SetConfiguration request to Ratis to remove the decommissioned peer from Raft conf and OM state.

If the force option is used in the decommission command,  the check that all OMs have the updated config will be bypassed and the decommission request will be sent to the OM leader.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-5490

## How was this patch tested?

Will update unit tests

bharatviswa504 reviewed (commented):
@hanishakoneru 
Thank You for the PR.
I have started looking in to this, can you add a high level explanation about decomission process and the flow. It will help in better clarity for review.

bharatviswa504 commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java:
```diff
@@ -1596,37 +1602,66 @@ public void bootstrap(OMNodeDetails newOMNode) throws IOException {
+        try {
+          removeOMNodeFromPeers(omNodeId);
+        } catch (IOException e) {
+          e.printStackTrace();
```
Do we want to log here?

hanishakoneru replied:
Yes, updated.

bharatviswa504 commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ratis/OzoneManagerRatisServer.java:
```diff
@@ -330,14 +330,45 @@ public void addOMToRatisRing(OMNodeDetails newOMNode) throws IOException {
+      throws IOException {
+    Preconditions.checkNotNull(removeOMNode);
+
+    String removeNodeId = removeOMNode.getNodeId();
```
Do we want to allow only operation at a time on this operation? And throw exception to other invokes on this API from client Like Already remove is ongoing something like this.

Coming from the perspective of one cli remove om1, and another cli remove om2. If we know inprogress thing, we can ask for 2nd invocation retry once after first is completed. Do we need any such semantics here?

bharatviswa504 replied:
And also one question how remove works when leader OM needs to be removed? 
And when remove in setConf we update peerlist, and add/remove grpcAppender Threads.

bharatviswa504 replied:
I think after reading test understood, we can stop and then decomission. So, we don't need to worry about that if leader is the one being removed i think. Is my understanding correct?

hanishakoneru replied:
With the current patch, we cannot remove leader OM. So yes, OM can be stopped before decommissioning. But if there are only 2 OMs in the ring, and we stop the follower to decommission, then the Leader will not be able to process the SetConf request as it won't get the quorum.
So, the more general rule would be that decommissioning OM should not be the leader.

JacksonYao287 replied:
if we want to remove a leader from the ratis ring, can we first transfer the leadership to a follower and then remove the old leader(now it is follower)

hanishakoneru replied:
@JacksonYao287, yes that should be possible. As long as a node is not currently the leader, it can be decommissioned.

bharatviswa504 commented on hadoop-ozone/tools/src/main/java/org/apache/hadoop/ozone/admin/om/DecommissionOMSubcommand.java:
```diff
@@ -0,0 +1,206 @@
+  private void verifyConfigUpdatedOnAllOMs() throws IOException {
+    String decommNodesKey = ConfUtils.addKeySuffixes(
+        OZONE_OM_DECOMMISSIONED_NODES_KEY, omServiceId);
+    Collection<String> decommNodes = ozoneConf.getTrimmedStringCollection(
```
This means before decomission, the config needs to be updated on all client/OMs. Is it correct?

hanishakoneru replied:
Yes, would have to update config on all OMs atleast before decommissioning.

bharatviswa504 commented on hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestAddRemoveOzoneManager.java:
```diff
@@ -333,4 +340,78 @@ public void testForceBootstrap() throws Exception {
+    cluster.stopOzoneManager(omNodeId2);
+
+    // Verify that we can read/ write to the cluster with only 1 OM.
+    OzoneVolume volume = objectStore.getVolume(VOLUME_NAME);
```
From above test we stopped 2 OM's how read/write will succeed?

hanishakoneru replied:
Once we remove 2 nodes, the ratis ring becomes a single node ring.

bharatviswa504 replied:
Got it. Now it makes sense why test passes :)

bharatviswa504 commented:
Thank You @hanishakoneru for the PR.
I have one general question.
With this approach, looks like we need the config to be updated before on all nodes not only OM right. As the decomission sub command also verifies few checks and also in general RpcClient also needs when building failOverProxy list.

hanishakoneru commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/protocol/OMConfiguration.java:
```diff
@@ -84,9 +84,21 @@ public OMConfiguration build() {
-        NodeDetails::getNodeId,
-        omNodeDetails -> omNodeDetails,
-        (nodeId, omNodeDetails) -> omNodeDetails));
+    return omNodesInNewConf.stream()
```
omNodesInNewConf include the decommissioned nodes. But we want only active nodes from getOmNodesInNewConf().
I will update the function name and JavaDoc to clarify.

hanishakoneru commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/protocolPB/OMAdminProtocolClientSideImpl.java:
```diff
@@ -123,12 +175,47 @@ public OMConfiguration getOMConfiguration() throws IOException {
+    }
+
+    if (!response.getSuccess()) {
+      throwException("Decommission Request to " + omPrintInfo +
```
Done.

hanishakoneru commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/protocolPB/OMAdminProtocolClientSideImpl.java:
```diff
@@ -86,16 +101,53 @@ public OMAdminProtocolClientSideImpl(ConfigurationSource conf,
+        new OMFailoverProxyProvider(conf, ugi, omServiceId,
+            OMAdminProtocolPB.class);
+
+    // Multiple the max number of retries with number of OMs to calculate the
```
Done.

hanishakoneru commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/protocolPB/OMAdminProtocolClientSideImpl.java:
```diff
@@ -86,16 +101,53 @@ public OMAdminProtocolClientSideImpl(ConfigurationSource conf,
+  /**
+   * Create OM Admin Protocol Client for contacting the OM ring (failover
+   * till the current OM leader is reached). Use for admin commands such as
+   * decommissionOM which are should reach the OM leader.
```
Removed.

hanishakoneru commented on hadoop-hdds/tools/src/main/java/org/apache/hadoop/hdds/cli/OzoneAdmin.java:
```diff
@@ -60,6 +65,16 @@ public OzoneConfiguration getOzoneConf() {
 
+  public UserGroupInformation getUser() throws IOException {
+    if (user == null) {
+      user = Server.getRemoteUser();
```
Updated to get current user.

hanishakoneru commented:
Thanks for the review @bharatviswa504. I have addressed your comments.

> With this approach, looks like we need the config to be updated before on all nodes not only OM right. As the decomission sub command also verifies few checks and also in general RpcClient also needs when building failOverProxy list.

Yes, its good to have the config updated on all nodes for failovers. But the requirement is only OMs and the node from which the admin command is run. RpcClient would still work with old configs (there would be a unnecessary failover to the decommissioned OM though).

bharatviswa504 commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/protocolPB/OMAdminProtocolClientSideImpl.java:
```diff
@@ -123,12 +175,48 @@ public OMConfiguration getOMConfiguration() throws IOException {
+
+    DecommissionOMResponse response;
+    try {
+      response = rpcProxy.decommission(NULL_RPC_CONTROLLER, decommOMRequest);
```
When reconfiguration exception, when already a in progress request, in failOver we need to retry on same OM for the ReconfigurationInProgressException

hanishakoneru replied:
Good catch Bharat.. I can update the OMFailoverProxyProvider#getRetryPolicy() to include this condition for ReconfigurationInProgressException. But this condition need not be checked for regular client requests.

So I was thinking we should improve the retry policy to be specific for each protocol instead of a generic one. There is a Jira (HDDS-5847) open for adding a separate failover proxy for OMInterServiceProtocol (ReconfigurationInProgressException is valid for inter service protocol too). I can update this along with that, if you are ok with that?

bharatviswa504 replied:
Ya sure. Pls update HDDS-5847 so that we don't miss it.

bharatviswa504 reviewed (commented):
Do we want to allow only operation at a time on this operation? And throw exception to other invokes on this API from client Like Already remove is ongoing something like this.

Coming from the perspective of one cli remove om1, and another cli remove om2. If we know inprogress thing, we can ask for 2nd invocation retry once after first is completed. Do we need any such semantics here?

Any comments on the above question. (From what i understand when in progress setconf, we get RaftReconfigurationException, and now we retry for configured retry count.)

Remaining PR overall LGTM, except one minor comment/question.

hanishakoneru commented:
Thanks for the review Bharat.

> Do we want to allow only operation at a time on this operation? And throw exception to other invokes on this API from client Like Already remove is ongoing something like this.
Coming from the perspective of one cli remove om1, and another cli remove om2. If we know inprogress thing, we can ask for 2nd invocation retry once after first is completed. Do we need any such semantics here?

Missed this earlier. Yes, it would be good to have this. But since the SetConfiguration will be synchronized from Ratis side, it is not critical that we control it from Ozone side as well. 
A simple way to handle this would be to fail the request on getting ReconfigurationInProgressException. This way we can throw the error upto the cli and ask to retry. What do you think?

bharatviswa504 commented:
> Thanks for the review Bharat.
> 
> > Do we want to allow only operation at a time on this operation? And throw exception to other invokes on this API from client Like Already remove is ongoing something like this.
> > Coming from the perspective of one cli remove om1, and another cli remove om2. If we know inprogress thing, we can ask for 2nd invocation retry once after first is completed. Do we need any such semantics here?
> 
> Missed this earlier. Yes, it would be good to have this. But since the SetConfiguration will be synchronized from Ratis side, it is not critical that we control it from Ozone side as well. A simple way to handle this would be to fail the request on getting ReconfigurationInProgressException. This way we can throw the error upto the cli and ask to retry. What do you think?

Yes good idea, i think this can also be taken as part of HDDS-5847, as we need changes in failOverProxyProvider for that to fail when ReconfigurationInProgressException is received.

bharatviswa504 reviewed (commented):
+1 LGTM
Based on the comments, that few of them will be revisited as part of HDDS-5847.

hanishakoneru commented:
Thanks @bharatviswa504 for the reviews and the +1. I have updated HDDS-5847 to include policy for ReconfigurationInProgressException.
I will merge this shortly.
